pkg/certificates/certificates.go

```diff
@@ -0,0 +1,270 @@
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/gob"
+	"fmt"
+
+	"github.com/btcsuite/btcutil/base58"
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	"gopkg.in/spacemonkeygo/monkit.v2"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/peertls"
+	"storj.io/storj/pkg/provider"
+	"storj.io/storj/pkg/utils"
+	"storj.io/storj/storage"
+	"storj.io/storj/storage/boltdb"
+	"storj.io/storj/storage/redis"
+)
+
+const (
+	// AuthorizationsBucket is the bucket used with a bolt-backed authorizations DB.
+	AuthorizationsBucket = "authorizations"
+	tokenLength          = 64 // 2^(64*8) =~ 1.34E+154
+)
+
+var (
+	mon = monkit.Package()
+	// ErrAuthorization is used when an error occurs involving an authorization.
+	ErrAuthorization = errs.Class("authorization error")
+	// ErrAuthorizationDB is used when an error occurs involving the authorization database.
+	ErrAuthorizationDB = errs.Class("authorization db error")
+	// ErrNotEnoughRandom is used when the number of bytes read from a random source are insufficient.
+	ErrNotEnoughRandom = ErrAuthorization.New("unable to read enough random bytes")
+	// ErrAuthorizationCount is used when attempting to create an invalid number of authorizations.
+	ErrAuthorizationCount = ErrAuthorizationDB.New("cannot add less than one authorizations")
+)
+
+// CertSignerConfig is a config struct for use with a certificate signing service
+type CertSignerConfig struct {
+	AuthorizationDBURL string `help:"url to the certificate signing authorization database" default:"bolt://$CONFDIR/authorizations.db"`
+}
+
+// CertificateSigner implements pb.CertificatesServer
+type CertificateSigner struct {
+	Log *zap.Logger
+}
+
+// AuthorizationDB stores authorizations which may be claimed in exchange for a
+// certificate signature.
+type AuthorizationDB struct {
+	DB storage.KeyValueStore
+}
+
+// Authorizations is a slice of authorizations for convenient de/serialization
+// and grouping.
+type Authorizations []*Authorization
+
+// Authorization represents a single-use authorization token and it's status
+type Authorization struct {
+	Token Token
+	Claim *Claim
+}
+
+// Token is a random byte array to be used like a pre-shared key for claiming
+// certificate signatures.
+type Token [tokenLength]byte
+
+// Claim holds information about the circumstances under which an authorization
+// token was claimed.
+type Claim struct {
+	IP         string
+	Timestamp  int64
+	Identity   *provider.PeerIdentity
+	SignedCert *x509.Certificate
+}
+
+// NewServer creates a new certificate signing grpc server
+func NewServer(log *zap.Logger) pb.CertificatesServer {
+	srv := CertificateSigner{
+		Log: log,
+	}
+
+	return &srv
+}
+
+// NewAuthorization creates a new, unclaimed authorization with a random token value
+func NewAuthorization() (*Authorization, error) {
+	var token Token
+	i, err := rand.Read(token[:])
+	if err != nil {
+		return nil, ErrAuthorization.Wrap(err)
+	}
+	if i != tokenLength {
+		return nil, ErrNotEnoughRandom
+	}
```

Comment:
is this a legit concern?

Reply:
From https://golang.org/pkg/crypto/rand/:
"Read is a helper function that calls Reader.Read using io.ReadFull. On return, n == len(b) if and only if err == nil. "

In other words, it does not seem like this is a concern since `i==tokenLength` if there is no error

Reply:
nice - thanks for checking... now I just feel lazy :p